libs/filer/workspace_fuse_client.go

```diff
@@ -0,0 +1,218 @@
+package filer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"path"
+	"slices"
+	"strings"
+
+	"github.com/databricks/cli/libs/notebook"
+	"github.com/databricks/databricks-sdk-go"
+	"github.com/databricks/databricks-sdk-go/service/workspace"
+	"golang.org/x/exp/maps"
+)
+
+type workspaceFuseClient struct {
+	workspaceFilesClient Filer
+}
+
+func stripNotebookExtension(ctx context.Context, w Filer, name string) (stripName string, stat fs.FileInfo, ok bool) {
+	ext := path.Ext(name)
+
+	extensionsToLanguages := map[string]workspace.Language{
+		".py":    workspace.LanguagePython,
+		".r":     workspace.LanguageR,
+		".scala": workspace.LanguageScala,
+		".sql":   workspace.LanguageSql,
+		".ipynb": workspace.LanguagePython,
+	}
+
+	stripName = strings.TrimSuffix(name, ext)
+
+	// File name does not have an extension associated with Databricks notebooks, return early.
+	if !slices.Contains(maps.Keys(extensionsToLanguages), ext) {
+		return "", nil, false
+	}
+
+	// If the file could be a notebook, check it's a notebook and has the correct language.
+	stat, err := w.Stat(ctx, stripName)
+	if err != nil {
+		return "", nil, false
+	}
+	info := stat.Sys().(workspace.ObjectInfo)
+
+	// Not a notebook. Return early.
+	if info.ObjectType != workspace.ObjectTypeNotebook {
+		return "", nil, false
+	}
+
+	// Not the correct language. Return early.
+	if info.Language != extensionsToLanguages[ext] {
+		return "", nil, false
+	}
+
+	return stripName, stat, true
+}
+
+type DuplicatePathError struct {
+	oi1 workspace.ObjectInfo
+	oi2 workspace.ObjectInfo
+
+	commonName string
+}
+
+func (e DuplicatePathError) Error() string {
+	return fmt.Sprintf("cannot read files from the workspace file system by emulating FUSE. Duplicate paths encountered. Both %s at %s and %s at %s resolve to the same name %s", e.oi1.ObjectType, e.oi1.Path, e.oi2.ObjectType, e.oi2.Path, e.commonName)
+}
+
+// This is a filer for the workspace file system that allows you to pretend the
+// workspace file system is a traditional file system. It allows you to list, read, write,
+// delete, and stat notebooks (and files in general) in the workspace, using their paths
+// with the extension included.
```

Comment:
This is a great summary. Also highlights why using the reference to FUSE is not applicable here. This is a workspace files with extension filer. It doesn't have anything to do with FUSE, we just hope that over time we can replace it with a local filesystem filer with the same semantics.

Reply:
Removed mentions of FUSE from the filer client.